core/src/main/java/com/google/errorprone/bugpatterns/nullness/RedundantNullCheck.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2025 The Error Prone Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.errorprone.bugpatterns.nullness;
+
+import static com.google.errorprone.BugPattern.SeverityLevel.WARNING;
+import static com.google.errorprone.matchers.Description.NO_MATCH;
+import static com.google.errorprone.matchers.method.MethodMatchers.staticMethod;
+import static com.google.errorprone.util.ASTHelpers.hasImplicitType;
+import static com.google.errorprone.util.ASTHelpers.isConsideredFinal;
+import static javax.lang.model.type.TypeKind.TYPEVAR;
+
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.ErrorProneFlags;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.bugpatterns.BugChecker.BinaryTreeMatcher;
+import com.google.errorprone.bugpatterns.BugChecker.MethodInvocationTreeMatcher;
+import com.google.errorprone.bugpatterns.nullness.NullnessUtils.NullCheck;
+import com.google.errorprone.dataflow.nullnesspropagation.Nullness;
+import com.google.errorprone.dataflow.nullnesspropagation.NullnessAnnotations;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.util.ASTHelpers;
+
+import com.sun.source.tree.BinaryTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.LiteralTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.VariableTree;
+import com.sun.tools.javac.code.Symbol;
+import com.sun.tools.javac.code.Symbol.MethodSymbol;
+import com.sun.tools.javac.code.Symbol.VarSymbol;
+
+import java.util.Optional;
+
+import javax.inject.Inject;
+import javax.lang.model.element.ElementKind;
+
+@BugPattern(
+    summary =
+        "Null check on an expression that is statically determined to be non-null according to "
+            + "language semantics or nullness annotations.",
+    severity = WARNING)
+public class RedundantNullCheck extends BugChecker
+    implements BinaryTreeMatcher, MethodInvocationTreeMatcher {
+
+  private static final Matcher<ExpressionTree> OBJECTS_REQUIRE_NON_NULL =
+      staticMethod().onClass("java.util.Objects").named("requireNonNull");
+
+  private final boolean checkRequireNonNull;
+
+  @Inject
+  public RedundantNullCheck(ErrorProneFlags flags) {
```

Comment:
```suggestion
  RedundantNullCheck(ErrorProneFlags flags) {
```

Just to follow our internal style of reducing the visibility of constructors that are going to be called only through reflection.